File: demo/hello.go
// Matcha syntax highlighting demo — Go

package main

import (
	"fmt"
	"math"
	"strings"
)

const maxRetries = 3
const version = "0.1.0"

// Point represents a 2D coordinate.
type Point struct {
	X float64
	Y float64
}

func (p Point) Distance(other Point) float64 {
	dx := p.X - other.X
	dy := p.Y - other.Y
	return math.Sqrt(dx*dx + dy*dy)
}

func (p Point) String() string {
	return fmt.Sprintf("(%g, %g)", p.X, p.Y)
}

// Shape is implemented by geometric types.
type Shape interface {
	Area() float64
	Perimeter() float64
}

type Circle struct {
	Center Point
	Radius float64
}

func (c Circle) Area() float64 {
	return math.Pi * c.Radius * c.Radius
}

func (c Circle) Perimeter() float64 {
	return 2 * math.Pi * c.Radius
}

type Rectangle struct {
	Origin Point
	Width  float64
	Height float64
}

func (r Rectangle) Area() float64 {
	return r.Width * r.Height
}

func (r Rectangle) Perimeter() float64 {
	return 2 * (r.Width + r.Height)
}

func fibonacci(n int) []int {
	if n <= 0 {
		return nil
	}
	fibs := make([]int, n)
	fibs[0] = 0
	if n > 1 {
		fibs[1] = 1
		for i := 2; i < n; i++ {
			fibs[i] = fibs[i-1] + fibs[i-2]
		}
	}
	return fibs
}

func main() {
	fmt.Printf("Matcha demo v%s\n", version)

	origin := Point{0, 0}
	p := Point{3, 4}
	fmt.Printf("Distance from %s to %s: %.2f\n", origin, p, origin.Distance(p))

	/* Multi-line block comment:
	   This demonstrates that block comments
	   are highlighted correctly across lines. */
	shapes := []Shape{
		Circle{Center: origin, Radius: 5},
		Rectangle{Origin: origin, Width: 10, Height: 20},
	}

	for _, shape := range shapes {
		fmt.Printf("Area: %.2f, Perimeter: %.2f\n", shape.Area(), shape.Perimeter())
	}

	fibs := fibonacci(10)
	parts := make([]string, len(fibs))
	for i, v := range fibs {
		parts[i] = fmt.Sprintf("%d", v)
	}
	fmt.Printf("Fibonacci: [%s]\n", strings.Join(parts, ", "))
}

File: demo/hello.rs
// Matcha syntax highlighting demo — Rust

use std::collections::HashMap;
use std::fmt;

const MAX_SIZE: usize = 256;
static GREETING: &str = "Hello from Matcha!";

#[derive(Debug, Clone)]
struct Point {
    x: f64,
    y: f64,
}

impl Point {
    fn new(x: f64, y: f64) -> Self {
        Point { x, y }
    }

    fn distance(&self, other: &Point) -> f64 {
        let dx = self.x - other.x;
        let dy = self.y - other.y;
        (dx * dx + dy * dy).sqrt()
    }
}

impl fmt::Display for Point {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({}, {})", self.x, self.y)
    }
}

#[derive(Debug)]
enum Shape {
    Circle { center: Point, radius: f64 },
    Rectangle { origin: Point, width: f64, height: f64 },
}

impl Shape {
    fn area(&self) -> f64 {
        match self {
            Shape::Circle { radius, .. } => std::f64::consts::PI * radius * radius,
            Shape::Rectangle { width, height, .. } => width * height,
        }
    }
}

fn fibonacci(n: u32) -> Vec<u64> {
    let mut fibs = vec![0u64, 1];
    for i in 2..n as usize {
        let next = fibs[i - 1] + fibs[i - 2];
        fibs.push(next);
    }
    fibs
}

fn main() {
    println!("{}", GREETING);

    let origin = Point::new(0.0, 0.0);
    let p = Point::new(3.0, 4.0);
    println!("Distance: {:.2}", origin.distance(&p));

    // Shapes
    let shapes: Vec<Shape> = vec![
        Shape::Circle { center: origin.clone(), radius: 5.0 },
        Shape::Rectangle { origin: origin.clone(), width: 10.0, height: 20.0 },
    ];

    for shape in &shapes {
        println!("{:?} -> area = {:.2}", shape, shape.area());
    }

    // HashMap
    let mut scores: HashMap<&str, i32> = HashMap::new();
    scores.insert("Alice", 95);
    scores.insert("Bob", 87);

    if let Some(&score) = scores.get("Alice") {
        println!("Alice scored {}", score);
    }

    let fibs = fibonacci(10);
    let sum: u64 = fibs.iter().sum();
    println!("Fibonacci sum: {}", sum);
}
